r/tests/testthat/test-Array.R

```diff
@@ -202,6 +202,8 @@ test_that("array supports Date (ARROW-3340)", {
   expect_array_roundtrip(d2, date32())
   # PSA: IngestDoubleRange(Date32Builder) truncates decimals, so this only
   # works where the dates are integer-ish
+
+  expect_equal(typeof(Array$create(d)$as_vector()), "double")
```

Comment:
If we want to assert about R storage type, I'd move the into `expect_array_roundtrip()` defined at the top of this file, below L30:

```r
expect_identical(typeof(as.vector(a)), typeof(x))
```

so that we can test all types. (I did this locally just now on master, and only Date fails.)

I also wonder if it would be good to add a test here with `dplyr::if_else` to reproduce the original issue. IMO the storage mode is only interesting because of the behavior side effects.